### PR TITLE
plutus-tx: add a list primitive to `Data`

### PR DESCRIPTION
**Pull request by michaelpj**
This seems like a good idea:
- CBOR supports it
- Lists are otherwise encoded pretty poorly

I also fixed up a TODO.

### CODE REVIEW

File: plutus-tx/compiler/Language/PlutusTx/Data.hs [hunk 1]
```diff
@@ -27,35 +28,53 @@ import           GHC.Generics
 data Data =
       Constr Integer [Data]
     | Map [(Data, Data)]
+    | Seq [Data]
```
Comment (kwxm) [inline plutus-tx/compiler/Language/PlutusTx/Data.hs]: Any particular reason it's called Seq and not Map?

This prompts me to wonder whether we need to care about the semantics of `Data`.  One interpretation of Seq is linked lists, but another could be random access arrays/vectors with constant access time.  I suppose it's up to the scripting language how it deals with these things.

**Comment by kwxm** [comment]
This all looks eminently sensible. Don't hesitate to add more of these if it looks as if they'd be useful.

File: plutus-tx/src/Language/PlutusTx/IsData.hs [hunk 2]
```diff
@@ -78,12 +79,9 @@ instance (IsData a, IsData b) => IsData (Either a b) where
     fromData (Constr i [bd]) | i == 1 = Left <$> fromData bd
     fromData _               = Nothing
 
--- TODO: maybe this should actually be encoded as a list!
 instance IsData a => IsData [a] where
     {-# INLINABLE toData #-}
-    toData [] = Constr 0 []
-    toData (x:xs) = Constr 1 [toData x, toData xs]
+    toData xs = Seq (fmap toData xs)
```
Comment (effectfully) [inline plutus-tx/src/Language/PlutusTx/IsData.hs]: It's kind of nice to have a separate constructor of `Data`, but I'd probably just define it as

```haskell
toData xs = Constr 0 $ map toData xs
```

(and `fromData` accordingly). That `Constr` is not self-describing when it comes to other data types (i.e. its actual meaning depends on the data type it's being converted to/from) and so I wouldn't make lists any more special in this respect.

Comment (effectfully) [reply plutus-tx/src/Language/PlutusTx/IsData.hs]: Although I admit it's kind of a hack.

Comment (michaelpj) [reply plutus-tx/src/Language/PlutusTx/IsData.hs]: Hm, the semantics are slightly off then. `Constr` is really supposed to have a tuple of arguments of some known size, so using it for variable length lists is a bit odd.

Note that it gets turned into almost the same thing in CBOR: a `Constr` is just a tagged list, a `Seq` is an untagged list. So the difference is mainly how it gets represented on the Plutus side.

File: plutus-tx/compiler/Language/PlutusTx/Data.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-tx/compiler/Language/PlutusTx/Data.hs]: Not particularly. I'll just call it `List`.

File: plutus-tx/src/Language/PlutusTx/IsData.hs (hunk 2 above)
Comment (effectfully) [reply plutus-tx/src/Language/PlutusTx/IsData.hs]: > Hm, the semantics are slightly off then. Constr is really supposed to have a tuple of arguments of some known size, so using it for variable length lists is a bit odd.

That's why I say it's a hack.

> Note that it gets turned into almost the same thing in CBOR: a Constr is just a tagged list, a Seq is an untagged list. 

Which is the reason I propose we can just use an empty tag.

Either way is fine, I just wanted to mention this option.

Comment (michaelpj) [reply plutus-tx/src/Language/PlutusTx/IsData.hs]: Yeah, it's not a bad idea I think I just like this one slightly more.

**Pull request merged**